xmtp/README.md

```diff
@@ -108,7 +127,50 @@ init():
     return
 ```
 
-### Receiving an invite
+### Receiving invites
+
+```plaintext
+downloadInvites():
+    Get the `refresh_jobs` record with an id of `invites`, and obtain a lock on the row:
+        Store `now()` in memory to mark the job execution start
+        Fetch all messages from invite topic with timestamp > refresh_job.last_run - PADDING_TIME # PADDING TIME accounts for eventual consistency of network. Maybe 30s.
+        For each message in topic:
+            Save (or ignore if already exists) raw message to inbound_invite table with status of PENDING
+        Update `refresh_jobs` record last_run = current_timestamp
+
+processInvites():
+    For each inbound_invite in PENDING state:
+        If an existing session exists with the `inviter.installation_id`:
+            Decrypt the inner invite using the existing session
+            If decryption fails:
+                Set inbound_invite state to DECRYPTION_FAILURE
+                continue
+            Update session in the database
+        else:
+            Create a new inbound session with the inviter
+            Decrypt the inner invite using the new session
+            If decryption fails:
+                Set inbound_invite state to DECRYPTION_FAILURE
+                continue
+            Persist the session to the database
```

Comment:
In the case there the decryption fails for a new session/PreKeyMessage, shouldn't the session be persisted regardless.

The session is valid outside of the invite, and subsequent messages from the remote_installation will continue to use the same session, albeit with increasing chain index values. By throwing away the session information does this not mean that all subsequent messages will need to be discarded even if they are valid?

Reply:
I don't think there is anything to save or update when decryption fails.

For creating a new inbound session, you need to decrypt the first inbound message as part of session creation. If that fails, there is no session to persist.

For decrypting a message using an existing session, if decryption fails the ratchet state would be unchanged so there is no need to update anything. The next message would just be treated as an out-of-order message from the perspective of the ratchet state.